### DISCUSSION TITLE
Is there a way to get results in other functions? A promise way? A subscribe way?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#1254] [Q&A] [answered]
I'm trying to understand how to get results in functions that I neither create nor control.

REPL: https://9crg7.csb.app/#/todos

Code example: https://codesandbox.io/s/urql-how-to-get-update-in-functions-9crg7?file=/List.svelte:266-816

```js
let condition = null;

function updateConditionAndgetTodosHere() {
  const unsubscribe = todos.subscribe((data) => {
    condition = { first: 1 };
    console.log("newData:", data);
    if (data.stale == false && data.fetching == false && data.todos) {
      console.log("data.todos:", data.todos);
      // unsubscribe(); // how to unsubscribe after all?
    }
  });
}

$: todos = query(
  operationStore(
    TODOS_QUERY,
    { input: condition },
    { requestPolicy: "cache-and-network", pause: !condition }
  )
);
```

I this example I need to `console.log(data.todos)`.

I can't figure out how to handle this.

Screen gif:

![issue](https://user-images.githubusercontent.com/41120635/103143624-55d9f000-471a-11eb-9da1-0ecd517d91a2.gif)

The only way I can get todos now is with `getClient()`, like this:

```js
const queryClient = getClient();

condition = { first: 1 };

const todos = await queryClient
  .query(
    TODOS_QUERY,
    { input: condition },
    // { requestPolicy: "cache-and-network" } // this doesn't work and I can understand why, but it's not good for me, I need cache-and-network here!
    { requestPolicy: "network-only" }
  )
  .toPromise();

if (todos.data) {
  // ok...
}
```

but I need `cache-and-network` which `getClient()` queries can't handle (obviously, obviously?).

How to do this? Where am I wrong?

**Comment by kitten** [comment, marked as answer]
Yes, using the client imperatively in any binding is recommended if the result is used outside of the normal flow of the UI, i.e. outside of the framework anyway.

I also before I get to a conclusion have to mention that overusing `cache-and-Network` often does more damage than what it's worth.

Any usage of an imperative method still goes through the usual flow of results, so suppose you have all three options of request policies availed, the usage of `cache-and-network` doesn't make sense, because there's nothing to update. You will likely always want to use `cache-first` or `network-only` imperatively.

When you use `cache-and-network` imperatively with a promise it obviously won't work since it will only receive the first result. When a query is active however in Svelte (or other bindings), that query will still receive the updated result after a while.

However, there's another reason why it doesn't make sense. The intention of why you have an imperative call is likely because you want to use the result after it's available. Hence, a `cache-and-netork` updated result wouldn't ever be available to you in an imperative flow, so it's the same as using `cache-first`. And if your function needs to use a fresh result `network-only` makes sense.

**Reply by frederikhors** [reply]
I would like to have a `cache-and-network` because that function is called with a `NEXT` button in a big todos table.

So I wanna display cache immediately and then update the table in background (if something is changed).

I think this makes sense too, right?

**Reply by frederikhors** [reply]
Can't we use the Svelte store's `subscribe()` for this?

What do you use in React for this scenario, guys?

**Reply by frederikhors** [reply]
Another example is a calendar, eg: https://fullcalendar.io/docs/events-function.

```js
events: function(info, successCallback, failureCallback) {
// here...
}
```

The calendar full of events...

**Reply by kitten** [reply]
I'm not sure the example is complete here. What do you mean by "display cache immediately"? As long as you still have an active query around it'll update appropriately

**Reply by frederikhors** [reply]
In the example of calendar, if I click Next button I can only use the `getClient()` way to query API.

Let's say the next page query (the next month query) is already cached in urql why I cannot get that cache and simultaneously check for new content in background?

The fullcalendar is an example of function I cannot control nor change.

I don't know if I've made myself clear. I'm better with REPLs maybe. :smile:

**Reply by frederikhors** [reply]
> As long as you still have an active query around it'll update appropriately

Not if I use `getClient()` `.toPromise()`. This is the problem.

**Reply by kitten** [reply]
So you mean suppose the query is on another page and you'd want to imperatively refetch it? You could send off `cache-first` to get the state and call a `network-only` operation manually as needed then.

But the point about `cache-and-network` not working out well with imperative logic still stands 😅

But that's partly because it's not too bad to reproduce the same logic with the promise-based API even.

So you won't need streaks even if you do something like this:

```js
const getResultAndUpdate = async () => {
  const refetch = client.query(...).toPromise();

  const result = client.readQuery(...) || (await refetch);

  // ...
};
```

where `refetch` calls `client.query` with `network-only`.

**Reply by frederikhors** [reply]
Wow. I didn't know about `client.readQuery()`. Thanks.

I updated the REPL here: https://codesandbox.io/s/urql-how-to-get-update-in-functions-forked-3sr58?file=/List.svelte

**Gif:**

![issue](https://user-images.githubusercontent.com/41120635/103213149-1aebde00-490d-11eb-891d-e2680eaf41c8.gif)

**Why is `result` always `null`? Even though I have cached elements:**

```js
const result = client.readQuery(TODOS_QUERY, { input: pagination });
```

Example:

```js
const client = getClient();

async function updateConditionAndgetTodosHere() {
  const pagination = { first: 1 };

  const todos = await getResultAndUpdate(pagination);

  if (todos.error) {
    console.log(todos.error.message);
  }

  if (todos.data) {
    console.log(todos.data);
  }
}

const getResultAndUpdate = async (pagination) => {
  const refetch = client
    .query(
      TODOS_QUERY,
      { input: pagination },
      { requestPolicy: "network-only" }
    )
    .toPromise();

  const result = client.readQuery(TODOS_QUERY, { input: pagination });
  console.log("result:", result); // This is always null! Why?

  return result || (await refetch);
};
```

Every time `refetch()` is called.

**Reply by frederikhors** [reply]
UPDATE: Gif added.

**Reply by JoviDeCroock** [reply]
This seems to be because `network-only` doesn't write to the cache, changing to `cache-and-network` did the trick for me.

Source: https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/cacheExchange.ts#L288

**Reply by frederikhors** [reply]
WOW! This works! 😃

Can I ask you to marry me, @kitten @JoviDeCroock? 😄

❤️❤️❤️

**Comment by frederikhors** [comment]
> What do you mean by "display cache immediately"?

Forgive me. In my REPL I would like to console.log() the cached todos and simultaneously fetch the new ones.

How can this be done today?
